### ISSUE TITLE
Issue with S3 Backup to OVH Object Storage - Failed to Upload Information Files

### ISSUE THREAD

**Issue by hamidamadani** [virtualmin/virtualmin-gpl#585]
Dear Virtualmin Team,

I am writing to report an issue I have encountered with the scheduled S3 backup to OVH object storage in Virtualmin. I have been utilizing this feature successfully, but starting from June 16, the backup process has been failing with the following error message:

"upload failed! Failed to upload information file: The request signature we calculated does not match the signature you provided. Check your key and signing method."

Upon investigation, I have found that while the actual backup files are being uploaded without any issues, the .info and .dom files are not being uploaded. Moreover, I noticed a warning displayed above the list of my S3 buckets, stating: "warning! the aws command is installed, but not working: could not connect to the endpoint URL: 'https://s3.sbg.amazonaws.com/'." It seems that in some pages, the user-specified endpoint is not getting used, probably including the code for uploading info files. Since the code used for uploading the actual backup files works fine, but uploading the info files doesn't.

This issue started occurring after installing the latest version (7.7) of Virtualmin, leading me to believe that it might be a new minor issue introduced in the update.

I want to emphasize that my S3 buckets are listed correctly, and their content is displayed properly. Only the uploading of the info files is affected. The backups themselves are being created without any problems.

I kindly request your assistance in resolving this matter. Any guidance or solution you can provide would be greatly appreciated. If possible, please consider investigating this issue and addressing it in an upcoming release or providing a workaround until a fix is available.

Thank you for your attention to this matter. I look forward to your response and resolution.

**Comment by jcameron** [comment]
Do you have the `aws`  CLI command installed on your system? Virtualmin uses this (if installed) to talk to all S3-compatible services ..

**Comment by hamidamadani** [comment]
Thank you for your response.

I confirm that I have the `aws CLI` command installed on my system, and I use it using the `--endpoint-url` parameter. So, you're saying that Virtualmin utilizes the `AWS CLI` without applying the user-specified endpoint. Instead, it uses the default Amazon endpoint only to upload information files, while the actual backup files are being uploaded in another way, which doesn't cause an issue. I appreciate the clarification provided. However, it's worth noting that this issue did not exist in previous versions of Virtualmin.

It appears that since there is no standard way to specify a custom endpoint URL in the `aws CLI` command, I'm curious if employing the `awscli-plugin-endpoint` plugin would resolve this problem by allowing me to specify a default endpoint for my `aws CLI`. Or I have to uninstall `aws CLI` command. Your insights on this matter would be greatly appreciated.

Once again, thank you for your response and assistance. Your time and consideration are greatly valued.

**Comment by jcameron** [comment]
Virtualmin should actually pass the specified endpoint to the `aws` command using that flag, if it's set and working.

That error message "aws command is installed, but not working: could not connect to the endpoint URL" is suspicious though! 

Can you post a screenshot of the settings on the Cloud Storage Provider page in Virtualmin for S3?

**Comment by hamidamadani** [comment]
Sure. Here you are:
![image](https://github.com/virtualmin/virtualmin-gpl/assets/47269359/dcae0fee-9a06-403d-b639-199b6743979c)
![image](https://github.com/virtualmin/virtualmin-gpl/assets/47269359/a9f57750-d62b-4733-a8da-898e3aa9f625)

**Comment by jcameron** [comment]
It's odd that you're seeing an error about an endpoint URL that contains amazonaws.com when the endpoint you've entered is a completely different domain!

I wonder if this is due to some misconfiguration in the `aws` command. What does `/root/.aws/config` contain on your system? In particular, the section of that config file that starts with the access key ID.

**Comment by iliajie** [comment]
Jamie, I have run tests on my side and I can see the only difference between what `aws configure` and Virtualmin produces in output for `~/.aws` configs is that `aws configure` adds to `~/.aws/config` the region if set, i.e.:

```
[default]
region = us-east-1
```

Virtualmin doesn't do it. If I assumed that Virtualmin later passes the region as command line parameter, would that be correct .. otherwise we're just not setting the region correctly?

@hamidamadani If you manually edit `~/.aws/config` file and set the region as in example above &mdash; does it work for you then?

**Comment by hamidamadani** [comment]
@jcameron Thanks. The content of `/root/.aws/config` is:
```
[profile 90ff34bd3e**********327924174e9e]
```

@iliajie Thanks for checking. Please note that my desired endpoint is `s3.sbg.io.cloud.ovh.net`. Should I set the region as follows?
```
[default]
region = us-east-1
```

**Comment by iliajie** [comment]
> Should I set the region as follows?

Yes, try that!

**Comment by hamidamadani** [comment]
@iliajie I tried that, so my /root/.aws/config file is as follows:

```
[default]
region = us-east-1

[profile 90ff34bd3e724cb5b269327924174e9e]

```

I didn't reboot after the edit, I don't think it is needed (Is it?).

The warning message above _Edit Cloud Storage Provider_ and _Amazon S3 Buckets_ page are still persists:
`Warning!  The AWS command is installed, but not working : Could not connect to the endpoint URL: "https://s3.sbg.amazonaws.com/"`

I also ran my backup schedule. I see no change in the result:
```
.
.
.
Creating TAR file of home directory ..
    .. done

    Uploading archive to Amazon's S3 service ..
    .. upload failed! Failed to upload information file : The request signature we calculated does not match the signature you provided. Check your key and signing method.

.. completed in 1 minutes, 4 seconds


    Saving Virtualmin configuration ..
    .. done
.
.
.
```

**Comment by jcameron** [comment]
If you SSH into your system as `root` and run : 

```
aws --endpoint-url=https://s3.sbg.io.cloud.ovh.net --profile=90f***************************9e s3 ls
```

what does it output?

**Comment by hamidamadani** [comment]
The output of the command is:
```
Could not connect to the endpoint URL: "https://s3.sbg.amazonaws.com/"
```

So, is the issue about `AWS CLI`?

**Comment by jcameron** [comment]
Yeah, I'd suggest upgrading the AWS CLI command. If it can't connect to the endpoint URL, Virtualmin will fall back to it's own built-in S3 client which is less reliable for 3rd-party providers.

**Comment by hamidamadani** [comment]
Thanks @jcameron, the issue resolved.

By upgrading `AWS CLI` command to 2.13, the warning message above in Virtualmin remained unchanged but the error in running the command changed to: 
```
An error occurred (AuthorizationHeaderMalformed) when calling the ListBuckets operation: The authorization header is malformed; the region 'nov' is wrong; expecting 'sbg'
```

To fix this I added `region = sbg` under `[profile 90ff***************************9e]` section in `/root/.aws/config` file. After making this change, the warnings disappeared, and the backup process is now working correctly.

Thanks again.

**Comment by jcameron** [comment]
Excellent! FYI, the next release of Virtualmin will allow the region to be entered manually (rather than selected from the AWS standard list) for other S3-compatible cloud providers.

**Comment by iliajie** [comment]
Jamie, is this necessary for the AWS CLI config to contain:

```
[default]
region = us-east-1
```

.. I have read threads where the absence of this line reportedly caused various issues.

**Comment by jcameron** [comment]
I don't think it's necessary, no. And as far as I know, Virtualmin doesn't add that.

**Comment by iliajie** [comment]
I read a few posts on StackOverflow where many users confirmed that adding mentioned block to the config solved the initial problem. Running `aws configure` also adds this block .. 

https://stackoverflow.com/a/40411174/1455661

However, if we _always_ add `--region` flag to the `aws` command, it should also be fine, right .. ? So, do we always add the `--region` flag to `aws` command?

**Comment by jcameron** [comment]
No, we don't pass that flag all the time, because potentially we need to support buckets that are in multiple regions.

But maybe I need to re-consider this ...

**Comment by jcameron** [comment]
I'll work on this an update this ticket when it's fixed

**Comment by hamidamadani** [comment]
@iliajie 
> However, if we always include the `--region` flag in the `aws` command, things should work fine, right? So, do we need to add the `--region` flag every time we use the `aws` command?

As I mentioned, I had the correct `--region` flag in the `AWS CLI` command, but there were issues with both the `AWS CLI` command and Virtualmin's backup.

In other words, when running the following command:
```
aws --endpoint-url=https://s3.sbg.io.cloud.ovh.net --profile=90f***************************9e s3 ls
```
I encountered the following errors until I added the `region` parameter in the `AWS CLI` config file:
```An error occurred (AuthorizationHeaderMalformed) when calling the ListBuckets operation: The authorization header is malformed; the region 'nov' is incorrect; expecting 'sbg'```

**Comment by jcameron** [comment]
I looked into this some more with OVH, using the endpoint URL s3.us-east-va.io.cloud.ovh.us which includes the region (us-east-va). And it worked fine even though Virtualmin didn't set the --region flag.

**Comment by iliajie** [comment]
Ref.: https://forum.virtualmin.com/t/testing-ovh-s3-object-storage/122347/4?u=ilia

**Comment by jcameron** [comment]
We can't always set that flag, because sometimes there is no region needed, and other times the request is global. Also it shouldn't be necessary if the region is set correctly in the config files under `.aws`

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.
